cranelift/codegen/src/machinst/mod.rs

```diff
@@ -144,6 +144,19 @@ pub trait MachInst: Clone + Debug {
     /// control flow.
     fn gen_jump(target: MachLabel) -> Self;
 
+    /// Generate a store of an immediate value to a register. Used by the
+    /// control plane to generate random instructions.
+    fn gen_imm(_value: u64, _dst: Writable<Reg>) -> Option<Self> {
+        None
+    }
+
+    /// Move a value from a general purpose register to an xmm one.
+    /// TODO this abstraction is kinda x64 specific...
+    /// but `gen_move` only works if the register types match?
+    fn gen_gpr_to_xmm(_src: Reg, _dst: Writable<Reg>) -> Option<Self> {
+        None
+    }
```

Comment:
An immediate floating point value seems to compile to the immediate value being stored in a general purpose register first, then moved to a floating point register (xmm).

So I did it the same way here. But I assume the abstraction is not right yet. The `MachInst` trait shouldn't contain references to implementation details of a specific backend.

If the `MachInst` trait had some method `gen_imm_f64`, what should it return? Just one of the existing instructions is not enough, I think, because a single instruction for immediate f64 values doesn't exist? And returning something like `Option<(Self, Self)>` also seems like a leaky abstraction.

@cfallin maybe you have an idea for the right abstraction here?

Reply:
@remlse a pattern we've used before successfully (see e.g. some instances in `abi.rs`) is to return a `SmallVec<[Inst; 2]>` (tweak inlined size as appropriate for expected max size). Then we have something like `for inst in M::gen_seq(...) { self.emit(inst); }` or whatever is appropriate at the callsite.

Reply:
That works well, thanks. `MachInst::gen_imm_f64` still "leaks" the fact that there may be a temporary general purpose register required, are you ok with that?

Reply:
Yep, that's fine, we have a notion of "this machine-specific sequence gets a tmp" elsewhere too. If you're happy with this please feel free to flip it to Ready to Review and I'll look over the whole thing!

Reply:
Ah, thanks, indeed that does not send an email so I missed it. Taking a final look now.